File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeSourceOperations.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.snowflake;

import static io.airbyte.cdk.db.jdbc.DateTimeConverter.putJavaSQLDate;
import static io.airbyte.cdk.db.jdbc.DateTimeConverter.putJavaSQLTime;

import com.fasterxml.jackson.databind.node.ObjectNode;
import io.airbyte.cdk.db.DataTypeUtils;
import io.airbyte.cdk.db.jdbc.JdbcSourceOperations;
import io.airbyte.commons.json.Jsons;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeFormatterBuilder;

public class SnowflakeSourceOperations extends JdbcSourceOperations {

  private static final DateTimeFormatter SNOWFLAKE_TIMESTAMPTZ_FORMATTER = new DateTimeFormatterBuilder()
      .parseCaseInsensitive()
      .append(DateTimeFormatter.ISO_LOCAL_DATE)
      .appendLiteral(' ')
      .append(DateTimeFormatter.ISO_LOCAL_TIME)
      .optionalStart()
      .appendLiteral(' ')
      .append(DateTimeFormatter.ofPattern("XX"))
      .toFormatter();

  @Override
  public void copyToJsonField(final ResultSet resultSet, final int colIndex, final ObjectNode json) throws SQLException {
    final String columnName = resultSet.getMetaData().getColumnName(colIndex);
    final String columnTypeName = resultSet.getMetaData().getColumnTypeName(colIndex).toLowerCase();

    switch (columnTypeName) {
      // jdbc converts VARIANT columns to serialized JSON, so we need to deserialize these.
      case "variant", "array", "object" -> json.set(columnName, Jsons.deserializeExact(resultSet.getString(colIndex)));
      default -> super.copyToJsonField(resultSet, colIndex, json);
    }
  }

  @Override
  protected void putDate(final ObjectNode node,
                         final String columnName,
                         final ResultSet resultSet,
                         final int index)
      throws SQLException {
    putJavaSQLDate(node, columnName, resultSet, index);
  }

  @Override
  protected void putTime(final ObjectNode node,
                         final String columnName,
                         final ResultSet resultSet,
                         final int index)
      throws SQLException {
    putJavaSQLTime(node, columnName, resultSet, index);
  }

  @Override
  protected void putTimestampWithTimezone(final ObjectNode node, final String columnName, final ResultSet resultSet, final int index)
      throws SQLException {
    final String timestampAsString = resultSet.getString(index);
    OffsetDateTime timestampWithOffset = OffsetDateTime.parse(timestampAsString, SNOWFLAKE_TIMESTAMPTZ_FORMATTER);
    node.put(columnName, timestampWithOffset.format(DataTypeUtils.TIMESTAMPTZ_FORMATTER));
  }

  protected void putTimestamp(final ObjectNode node, final String columnName, final ResultSet resultSet, final int index) throws SQLException {
    // for backward compatibility
    var instant = resultSet.getTimestamp(index).toInstant();
    node.put(columnName, DataTypeUtils.toISO8601StringWithMicroseconds(instant));
  }

}

File: airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/SnowflakeSourceOperations.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.integrations.destination.snowflake

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import io.airbyte.cdk.db.DataTypeUtils
import io.airbyte.cdk.db.DataTypeUtils.toISO8601StringWithMicroseconds
import io.airbyte.cdk.db.jdbc.DateTimeConverter.putJavaSQLDate
import io.airbyte.cdk.db.jdbc.DateTimeConverter.putJavaSQLTime
import io.airbyte.cdk.db.jdbc.JdbcSourceOperations
import io.airbyte.commons.json.Jsons.deserializeExact
import java.sql.ResultSet
import java.sql.SQLException
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeFormatterBuilder
import java.util.*

class SnowflakeSourceOperations : JdbcSourceOperations() {
    @Throws(SQLException::class)
    override fun copyToJsonField(resultSet: ResultSet, colIndex: Int, json: ObjectNode) {
        val columnName = resultSet.metaData.getColumnName(colIndex)
        val columnTypeName =
            resultSet.metaData.getColumnTypeName(colIndex).lowercase(Locale.getDefault())

        when (columnTypeName) {
            "variant",
            "array",
            "object" ->
                json.set<JsonNode>(columnName, deserializeExact(resultSet.getString(colIndex)))
            else -> super.copyToJsonField(resultSet, colIndex, json)
        }
    }

    @Throws(SQLException::class)
    override fun putDate(node: ObjectNode, columnName: String?, resultSet: ResultSet, index: Int) {
        putJavaSQLDate(node, columnName, resultSet, index)
    }

    @Throws(SQLException::class)
    override fun putTime(node: ObjectNode, columnName: String?, resultSet: ResultSet, index: Int) {
        putJavaSQLTime(node, columnName, resultSet, index)
    }

    @Throws(SQLException::class)
    override fun putTimestampWithTimezone(
        node: ObjectNode,
        columnName: String?,
        resultSet: ResultSet,
        index: Int
    ) {
        val timestampAsString = resultSet.getString(index)
        val timestampWithOffset =
            OffsetDateTime.parse(timestampAsString, SNOWFLAKE_TIMESTAMPTZ_FORMATTER)
        node.put(columnName, timestampWithOffset.format(DataTypeUtils.TIMESTAMPTZ_FORMATTER))
    }

    @Throws(SQLException::class)
    override fun putTimestamp(
        node: ObjectNode,
        columnName: String?,
        resultSet: ResultSet,
        index: Int
    ) {
        // for backward compatibility
        val instant = resultSet.getTimestamp(index).toInstant()
        node.put(columnName, toISO8601StringWithMicroseconds(instant))
    }

    companion object {
        private val SNOWFLAKE_TIMESTAMPTZ_FORMATTER: DateTimeFormatter =
            DateTimeFormatterBuilder()
                .parseCaseInsensitive()
                .append(DateTimeFormatter.ISO_LOCAL_DATE)
                .appendLiteral(' ')
                .append(DateTimeFormatter.ISO_LOCAL_TIME)
                .optionalStart()
                .appendLiteral(' ')
                .append(DateTimeFormatter.ofPattern("XX"))
                .toFormatter()
    }
}
